src/Main.elm

```diff
@@ -1,10 +1,396 @@
+        missingFromAudiences =
+            errors.missingFolderId
+                |> List.map
+                    (\audience ->
+                        Html.div [ errorClass ]
+                            [ Html.text <|
+                                "The audience \""
+                                    ++ audience.name
+                                    ++ "\" (id: "
+                                    ++ String.fromInt audience.id
+                                    ++ ") references "
+                                    ++ String.fromInt audience.parent
+                                    ++ " as folder id, but this folder doesn't exist. "
+                                    ++ " This audience has been added to the root."
+                            ]
+                    )
+
+        missingFromFolders =
+            errors.missingFolderParentId
+                |> List.map
+                    (\folder ->
+                        Html.div [ errorClass ]
+                            [ Html.text <|
+                                "The folder \""
+                                    ++ folder.name
+                                    ++ "\" (id: "
+                                    ++ String.fromInt folder.id
+                                    ++ ") references "
+                                    ++ String.fromInt folder.parent
+                                    ++ " as parent folder id, but this folder doesn't exist. "
+                                    ++ " This folder has been added to the root."
+                            ]
+                    )
+    in
+    Html.div
+        [ Html.Attributes.class "flex flex-col space-y-2" ]
+        (duplicated
+            ++ missingFromAudiences
+            ++ missingFromFolders
+        )
+
+
+errorClass : Html.Attribute msg
+errorClass =
+    Html.Attributes.class "p-5 w-64 rounded-md flex flex-row bg-red-800 text-red-100"
+
+
+
+-----------
+-- BREADCRUMBS
+-----------
+
+
+viewBreadcrumbs : LoadedModel -> Html LoadedMsg
+viewBreadcrumbs loadedModel =
+    toSplittedBreadcrumbs loadedModel.breadcrumbs
+        |> List.reverse
+        |> List.map
+            (\splitted ->
+                Html.button
+                    [ Html.Events.onClick (BreadcrumbClicked splitted)
+                    , Html.Attributes.class "text-blue-500 underline"
+                    ]
+                    [ Html.text (Data.AudienceTree.name splitted.currentTree) ]
+            )
+        |> (\breadcrumbsHtml ->
+                breadcrumbsHtml
+                    ++ [ Html.div []
+                            [ Html.text (Data.AudienceTree.name loadedModel.currentTree) ]
+                       ]
+           )
+        |> List.intersperse (Html.div [ Html.Attributes.class "mx-1" ] [ Html.text ">" ])
+        |> Html.div [ Html.Attributes.class "flex flex-row text-sm items-center" ]
+
+
+type alias SplittedBreadcrumb =
+    { currentTree : AudienceTree
+    , breadcrumbs : List AudienceTree
```

Comment:
Does this hold large parts of the whole tree in memory?
eg.
```elm
-- full tree = degenerate 1-child tree looking like [A,B,C,D,E,F,G,H]
{ currentTree = [D,E,F,G,H]
, breadcrumbs =
  [ [C,D,E,F,G,H]
  , [B,C,D,E,F,G,H]
  , [A,B,C,D,E,F,G,H]
  ]
}
```
What would you consider as alternatives if there was so much data in the full tree that this duplication started being a problem?

Reply:
Yeah, immutable structures are very cool concerning data sharing!!

However, it might be an issue if the folder structureis "linear" with a very long chain of folders "folder1 > folder2 > folder 3 >... > folder1000" since the complexity of building these breadcrumbs is quadratic in the depth of the tree. 

A possible solution would be to limit the breadcrumbs length to say 10 (which already seems a huge depth for a folder structure so I'm not sure I'd do that).